### PR TITLE
Add new functions to the drakvuf API for resolving full path of windows registry keys

### PR DESCRIPTION
**Pull request by aoshiken**
The new functions are:
- _drakvuf_reg_keybody_path()_: Returns path from CM_KEY_CONTROL_BLOCK pointer
- _drakvuf_reg_keycontrolblock_path()_: Returns allocated registry key full path from key object
- _drakvuf_reg_keyhandle_path()_: Returns an allocated registry key full path from key handle

Those functions are included in the new file `src/libdrakvuf/win-registry.c`

### WARNING
I tested this code **only** on a Windows 7 SP1 x64 platform although the code **probably** works on x86 too.

### USAGE
I'm assuming that the mainly used function will be _drakvuf_reg_keyhandle_path()_ (although I myself currently use the other two from some hooks deep down in the kernel).

The following example shows its use from a trap callback for the syscall _NtDeleteValueKey_, you can also use the same callback body for the syscalls _NtSetValueKey_ and _NtDeleteKey_ because all of them have a KEY HANDLE as the 1st parameter (we don't care about the value name):
```
static event_response_t hook_deletevaluekey_cb( drakvuf_t drakvuf, drakvuf_trap_info_t *info )
{
    if ( info->regs->rcx )
    {
        vmi_instance_t vmi = drakvuf_lock_and_get_vmi( drakvuf );

        char *key_path = drakvuf_reg_keyhandle_path( drakvuf, info, info->regs->rcx, 0 );

        if ( key_path )
        {
            printf("DELETEVALUEKEY: %s\n", key_path );
            free( key_path );
        }

        drakvuf_release_vmi( drakvuf );
    }
    return 0 ;
}
```

And below you can see the DRAKVUF output (sorted and snipped) with that plugin enabled and starting an Internet Explorer and trigger some manual actions from `regedit.exe`:
```
DELETEKEY: \REGISTRY\USER\S-1-5-21-2200285794-3943342076-1391323334-1000\SOFTWARE\ALFREDO1
DELETEKEY: \REGISTRY\USER\S-1-5-21-2200285794-3943342076-1391323334-1000\SOFTWARE\MICROSOFT\WINDOWS\CURRENTVERSION\INTERNET SETTINGS\5.0\LOWCACHE\EXTENSIBLE CACHE\MSHIST012017100420171005
DELETEVALUEKEY: \REGISTRY\USER\S-1-5-21-2200285794-3943342076-1391323334-1000\SOFTWARE\MICROSOFT\WINDOWS\CURRENTVERSION\INTERNET SETTINGS
DELETEVALUEKEY: \REGISTRY\USER\S-1-5-21-2200285794-3943342076-1391323334-1000\SOFTWARE\MICROSOFT\WINDOWS\CURRENTVERSION\INTERNET SETTINGS\ZONEMAP
DELETEVALUEKEY: \REGISTRY\USER\S-1-5-21-2200285794-3943342076-1391323334-1000\SOFTWARE\NEW KEY #4
SETVALUEKEY: \REGISTRY\USER\S-1-5-21-2200285794-3943342076-1391323334-1000\SOFTWARE\MICROSOFT\WINDOWS\CURRENTVERSION\INTERNET SETTINGS\ZONEMAP
SETVALUEKEY: \REGISTRY\USER\S-1-5-21-2200285794-3943342076-1391323334-1000\SOFTWARE\MICROSOFT\WINDOWS\CURRENTVERSION\INTERNET SETTINGS\ZONES
SETVALUEKEY: \REGISTRY\USER\S-1-5-21-2200285794-3943342076-1391323334-1000\SOFTWARE\NEW KEY #4
```
Hope this helps,

Regards

### CODE REVIEW

**Comment by tklengyel** [comment]
This is great, much appreciated!  =)

**Comment by tklengyel** [comment]
This is OK to test.

**Comment by tklengyel** [comment]
Also, would you mind adding a plugin that exercises this new API (ie. one that traps `NtDeleteValueKey`, `NtSetValueKey` and `NtDeleteKey`)?

**Comment by aoshiken** [comment]
Of course!! Which name will you prefere?

**Comment by aoshiken** [comment]
regmon? I'm afraid it's too much predictable... xD

**Comment by tklengyel** [comment]
Sure, regmon or registrymon would both be fine :)

**Comment by aoshiken** [comment]
Ok, I see, the error comes for not using the variable 'vmi', surprisingly I don't have this in my Debian Jessie.

**Comment by tklengyel** [comment]
Yes, the compile error comes from not actually using `vmi`. Locking vmi doesn't have any effect on the libdrakvuf use of the vmi instance, so you really only need to do that if you need access to it to do reading/writing. The reason why the error might not appear on other machines is because only the debug build of DRAKVUF enables `-Werror -Wall`.

File: src/libdrakvuf/win-registry.c [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+    char *buf_ret ;
+    vmi_instance_t vmi = drakvuf->vmi;
+    access_context_t ctx = {
+        .addr = p_key_control_block + drakvuf->offsets[ CM_KEY_NAMEBLOCK ],
+        .translate_mechanism = VMI_TM_PROCESS_DTB,
+        .dtb = info->regs->cr3,
+    };
+
+    if ( drakvuf->pm == VMI_PM_IA32E )
+        vmi_status = vmi_read_64( vmi, &ctx, &p_name_control_block );
```
[the remark this answers is not in the document]
Comment (aoshiken) [reply src/libdrakvuf/win-registry.c]: Done

File: src/libdrakvuf/win-registry.c [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    {
+        uint16_t name_length = 0 ;
+
+        ctx.addr = p_name_control_block + drakvuf->offsets[ CM_KEY_NAMELENGTH ] ;
+
+        if ( vmi_read_16( vmi, &ctx, &name_length ) == VMI_SUCCESS )
+        {
+            if ( name_length && ( name_length < 240 ) )
+            {
+                buf_ret = (char *)malloc( name_length + 1 );
```
[the remark this answers is not in the document]
Comment (aoshiken) [reply src/libdrakvuf/win-registry.c]: Done

File: src/libdrakvuf/win-registry.c [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+    vmi_instance_t vmi = drakvuf->vmi;
+    addr_t p_key_control_block = 0 ;
+    access_context_t ctx = {
+        .addr = p_key_body + drakvuf->offsets[ CM_KEY_CONTROL_BLOCK ],
+        .translate_mechanism = VMI_TM_PROCESS_DTB,
+        .dtb = info->regs->cr3,
+    };
+
+    if ( drakvuf->pm == VMI_PM_IA32E )
+        vmi_status = vmi_read_64( vmi, &ctx, &p_key_control_block );
```
[the remark this answers is not in the document]
Comment (aoshiken) [reply src/libdrakvuf/win-registry.c]: Done

File: src/libdrakvuf/win-registry.c [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+                vmi_status = vmi_read_64( vmi, &ctx, &p_key_control_block );
+            else
+                vmi_status = vmi_read_32( vmi, &ctx, (uint32_t *)&p_key_control_block );
+        }
+
+        if ( tot_len )
+        {
+            tot_len += g_slist_length( key_path_list ) + 1 ;
+
+            buf_ret = (char *)malloc( tot_len ) ;
```
[the remark this answers is not in the document]
Comment (aoshiken) [reply src/libdrakvuf/win-registry.c]: Done

File: src/libdrakvuf/win-registry.h [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef WIN_REGISTRY_H
```
[the remark this answers is not in the document]
Comment (aoshiken) [reply src/libdrakvuf/win-registry.h]: I know, I was just mimicing what the other "subsystems" (win-process, win-handles...) do. Anyway I'm getting rid of that header file

File: src/libdrakvuf/win-registry.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (aoshiken) [reply src/libdrakvuf/win-registry.c]: I'm sorry! I forgot that one... :)
Done!

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please.

**Comment by tklengyel** [comment]
Everything looks good, thanks again! ;)

**Comment by aoshiken** [comment]
Humm, I'm afraid I'm lost with that `"Closed with unmerged commits"`.

All of my commits have been against the same branch and merged perfectly while this PR was open so I don't understand that message. :?

Do you want me to open a new PR?

**Comment by tklengyel** [comment]
Not necessary, I just clicked on close instead of merge :) Once the CI gives the green light I can merge ;)

**Pull request merged**